torch/jit/__init__.py

```diff
@@ -379,7 +379,7 @@ def clone_input(a):
             return None
         elif isinstance(a, torch.Tensor):
             # TODO: figure out one liner to .clone() and set requires_grad
-            v = Variable(a.data.clone(memory_format=torch.preserve_format), requires_grad=a.requires_grad)
+            v = a.detach().clone(memory_format=torch.preserve_format).requires_grad_(a.requires_grad)
```

Comment:
for these `detach().clone(...)` calls -- why not put them in a no_grad block?  It seems generally preferable to have this order of preference:
1) no_grad
2) detach
3) .data

because you don't actually create this intermediate Tensor (`a.detach()`) that is potentially floating around.

Reply:
The version with no_grad does not look much clearer here I think:
```python
with torch.no_grad():
  v = a.clone(memory_format=torch.preserve_format)
v.requires_grad_(a.requires_grad)
```
The one liner above is clearer about what we want to do. And the intermediate Tensor is quite cheap.

Reply:
can you not set requires_grad under a no_grad block?  That doesn't seem right.

Reply:
As discussed offline, you can. But it is trickier as the user need to be aware if the operation returns a new Tensor or not. Will keep it as is for now.